### Limpieza del dataset Police_Arrests_20240702
En este notebook "01_limpeza" se ejecutan codigos para limpiar en función de los hallazgos del notebook anterio "00_exploracion".  

**Se buscará**
- `Limpiar valores en blanco`
- `Filtrar edades`
- `Agregar nuevas variables de interes en función de las existentes`

In [57]:
# Importo librerias de python
import pandas as pd
import matplotlib.pyplot as plt

In [58]:
# Declaro el dataframe a usar
eccomerce_sales =r'C:\Users\Administrador\OneDrive - udd.cl\Python para la ciencia de datos\proyecto_final\Data\Police_Arrests_20240702.csv'
df = pd.read_csv(eccomerce_sales)

In [59]:
# Tipo de variables iniciales del dataframe y valores nulos
info = df.info()
print(info)


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23510 entries, 0 to 23509
Data columns (total 10 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   ID Reference Number     23510 non-null  int64  
 1   Subject's race          23510 non-null  object 
 2   Subject's gender        23510 non-null  object 
 3   Subject's age           23510 non-null  int64  
 4   Ethnicity               23510 non-null  object 
 5   District of occurrence  23408 non-null  float64
 6   Adjacent to School      23510 non-null  int64  
 7   Assigned Division       23381 non-null  object 
 8   Assigned Bureau         18851 non-null  object 
 9   Event Date/Time         23510 non-null  object 
dtypes: float64(1), int64(3), object(6)
memory usage: 1.8+ MB
None


In [60]:
# Elimino filas con valores en blanco
cantidad_inicial = len(df)
print(f"El dataframe inicial tiene {cantidad_inicial} filas.")
df_sin_nulos = df.dropna()
cantidad_df_sin_nulos = len(df_sin_nulos)
cantidad_nulos = cantidad_inicial - cantidad_df_sin_nulos
print(f"Se filtraron {cantidad_nulos} filas por concepto de nulos.")
# Filtro edades menor a 80 y mayor a 18
cantidad_inicial_2 = len(df_sin_nulos)
df_limpio = df_sin_nulos.loc[(df_sin_nulos["Subject's age"] <= 80) & (df_sin_nulos["Subject's age"] >= 18)]
cantidad_filtrada = len(df_limpio)
cantidad_df_limpio = cantidad_inicial_2 - cantidad_filtrada
print(f"Se filtraron {cantidad_df_limpio} filas por concepto de edad.")
print(f"El nuevo dataframe tiene {cantidad_filtrada} filas.")

#Verifico nuevo "n" del dataframe
info = df_limpio.info()
print(info)

# Creo histograma para verificar cambios en edad
plt.hist(df_limpio["Subject's age"], bins=df['ID Reference Number '].nunique(), edgecolor='black')
plt.xlabel('Edad')
plt.ylabel('Frecuencia')
plt.title('Histograma de Edades')
plt.grid(True)
plt.show()

# Verifico nuevos estadísticos de edad
age_stats = df_limpio["Subject's age"].describe()
print(age_stats)



El dataframe inicial tiene 23510 filas.
Se filtraron 4735 filas por concepto de nulos.
Se filtraron 2458 filas por concepto de edad.
El nuevo dataframe tiene 16317 filas.
<class 'pandas.core.frame.DataFrame'>
Index: 16317 entries, 0 to 23509
Data columns (total 10 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   ID Reference Number     16317 non-null  int64  
 1   Subject's race          16317 non-null  object 
 2   Subject's gender        16317 non-null  object 
 3   Subject's age           16317 non-null  int64  
 4   Ethnicity               16317 non-null  object 
 5   District of occurrence  16317 non-null  float64
 6   Adjacent to School      16317 non-null  int64  
 7   Assigned Division       16317 non-null  object 
 8   Assigned Bureau         16317 non-null  object 
 9   Event Date/Time         16317 non-null  object 
dtypes: float64(1), int64(3), object(6)
memory usage: 1.4+ MB
None


**Limpieza inicial realizada**  
Se eliminaron valores en blanco, se filtro la variable correspondiente a la edad del sujeto por valores atipicos,por lo cual, se verificaron nuevos estadisticos de la variable filtrada "Subject's age"


**Filtrado de datos:**  
Se detalla el balance de datos filtrados:

| Cantidad de datos      | N             |
|------------------------|---------------|
| Dataframe Inicial      | 23.510        |
| Filtrados              | 7.193         |
| Dataframe limpio       | 16.317        |

**Cambios en variable númerica "Subject's age":**  
| Estadística | DF Inicial| DF Limpio   |
|-------------|-----------|-------------|
| count       | 23510     | 16317       |
| mean        | 31.813143 | 34.083287   |
| std         | 12.946909 | 12.139814   |
| min         | 1         | 18          | 
| 25%         | 22        | 25          |
| 50%         | 30        | 32          |
| 75%         | 39        | 41          |
| max         | 99        | 80          |


In [ ]:
# Modifico tipo de variable de "Event Date/Time"
df_limpio['Event Date/Time'] = pd.to_datetime(df_filtered['Event Date/Time'])
# Agrego columnas para año, mes, semana y día de la semana
df_limpio['Year'] = df_limpio['Event Date/Time'].dt.year
df_limpio['Month'] = df_limpio['Event Date/Time'].dt.month
df_limpio['Week'] = df_limpio['Event Date/Time'].dt.isocalendar().week
df_limpio['Weekday'] = df_limpio['Event Date/Time'].dt.day_name()

C:\Users\Administrador\AppData\Local\Temp\ipykernel_4956\3206749606.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_limpio['Event Date/Time'] = pd.to_datetime(df_filtered['Event Date/Time'])
C:\Users\Administrador\AppData\Local\Temp\ipykernel_4956\3206749606.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_limpio['Year'] = df_limpio['Event Date/Time'].dt.year
C:\Users\Administrador\AppData\Local\Temp\ipykernel_4956\3206749606.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy 

In [ ]:
# Creo una nueva columna 'AgeGroup' con las categorías de edad en función de "Subject's age"
bins = [18, 30, 50, float('inf')]
labels = ['Adulto joven', 'Adulto maduro', 'Adulto mayor']
df_limpio['AgeGroup'] = pd.cut(df_limpio["Subject's age"], bins=bins, labels=labels, right=False)

# Verifico agregación de variables al df_limpio
info = df_limpio.info()
print(info)

<class 'pandas.core.frame.DataFrame'>
Index: 16317 entries, 0 to 23509
Data columns (total 15 columns):
 #   Column                  Non-Null Count  Dtype         
---  ------                  --------------  -----         
 0   ID Reference Number     16317 non-null  int64         
 1   Subject's race          16317 non-null  object        
 2   Subject's gender        16317 non-null  object        
 3   Subject's age           16317 non-null  int64         
 4   Ethnicity               16317 non-null  object        
 5   District of occurrence  16317 non-null  float64       
 6   Adjacent to School      16317 non-null  int64         
 7   Assigned Division       16317 non-null  object        
 8   Assigned Bureau         16317 non-null  object        
 9   Event Date/Time         16317 non-null  datetime64[ns]
 10  Year                    16317 non-null  int32         
 11  Month                   16317 non-null  int32         
 12  Week                    16317 non-null  UInt32     

C:\Users\Administrador\AppData\Local\Temp\ipykernel_4956\1807298833.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_limpio['AgeGroup'] = pd.cut(df_limpio["Subject's age"], bins=bins, labels=labels, right=False)


In [ ]:
# Vertifico los posibles valores para las variables nuevas variables agredas
variables = ['Year', 'Month',
             'Week', 'Weekday', 'AgeGroup']


for var in variables:
    valores_unicos = df_filtered[var].unique()
    cantidad_valores = len(valores_unicos)
    print(f"- Los {cantidad_valores} posibles valores para {var} son: {valores_unicos}")

- Los 4 posibles valores para Year son: [2021 2022 2023 2024]
- Los 12 posibles valores para Month son: [ 1  2  3  4  5  6  7  8  9 10 11 12]
- Los 53 posibles valores para Week son: <IntegerArray>
[53,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17, 18,
 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37,
 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52]
Length: 53, dtype: UInt32
- Los 7 posibles valores para Weekday son: ['Friday' 'Saturday' 'Sunday' 'Monday' 'Tuesday' 'Wednesday' 'Thursday']
- Los 3 posibles valores para AgeGroup son: ['Adulto maduro', 'Adulto joven', 'Adulto mayor']
Categories (3, object): ['Adulto joven' < 'Adulto maduro' < 'Adulto mayor']


**Descripción del nuevas incorporadas al df_limpio**

- `1.-Year`: es el año del suceso basado en Event Date/Time  
- `2.-Month`: es el mes del suceso basado en Event Date/Time
- `3.-Week`: es la semana del suceso basado en Event Date/Time
- `4.-Week Day`: es el día de la semana del suceso basado en Event Date/Time
- `5.-'AgeGroup'`: Corresponde a una categorización de edad en tres niveles

**Finalizamos la limpieza del df**